00-programming-fundamentals/debugging-with-vs-code.md

```diff
@@ -0,0 +1,177 @@
+    end
+end
+
+#adds two numbers
+def add_them(a, b)
+    return a + b
+end
+
+#subtracts b from a
+def subtract_them(a, b)
+    return a - b
+end
+
+
+left_num = 5
+right_num = 3
+
+sum = cool_sum_func(left_num, right_num)
+
+puts sum
+```
+
+This is some very dopey code, but it will show us a lot about how the debugger gives us information.
+
+## Get Relevant Gems and Extensions
+If you have not already run the following:
+
+```bash
+$:gem install ruby-debug-ide
+$:gem install debase
+```
+
+Please run them now.
+
+## Configuring the Debugger
+
+VS Code has a very bare-bones debugger that comes with it, but it doesn't know how to talk to every language out of the box. When we set up projects in VS Code, we create a series of configurations that VS Code stores in a file called `launch.json` in a folder called `.vscode`.
+
+Open the debugger by using `⇧⌘D`, and look near the top left of the viewport to find a green triangle button that likely reads "No Configurations". Click on the text and select "Add Configuration" from the drop down menu. It doesn't matter what you pick after this, as we will be filling in the file for ourselves from this point forward. When you have finished selecting a pre-generated option, a new pane should open called `launch.json`. Replace the file with the following:
+
+```json
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "Debug Local File",
+            "type": "Ruby",
+            "request": "launch",
+            "cwd": "${workspaceRoot}",
+            "program": "${file}"
+        }
+    ]
+}
+```
+
+Once you have done so (and saved the `launch.json`) you will be able to select "Debug Local File" from the drop down. Additionally, when you go to add configurations in the future, it may show up as an option after selecting to debug in Ruby. 
+
+## Setting A Breakpoint 
+
+Now that we have the debugger prepped, we need to add a `breakpoint`. Breakpoints are like stop-signs alongside your code; they aren't written into the text of the code itself, but they tell the program to halt before continuing to the next command.
+
+We add them in VS Code by clicking on a line number that we want to stop on. In our current file, lines 4 and 7 are potentially interesting to us. Click to the left of the line number, or get the cursor to the line and use `f9` to add the breakpoint. When you've finished, you should see 2 red circles near the code, and the text next to the green play button should read "Debug Local File", see below:
+
+![In the editor, lines 4 and seven have bright red circles in front, indicating that have breakpoints set](images/breakpoints_vs_code.png "In the editor, lines 4 and seven have bright red circles in front, indicating that have breakpoints set")
```

Comment:
Hm, this image isn't showing up on GitHub for me, and I think it's because GitHub is being lovingly specific about the path capitalization: "images/breakpoints_vs_code.png" -> "images/breakpoints_VS_Code.png"